File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode fast = head , slow = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode temp = null;
        while(slow != null){
            ListNode nextNode = slow.next;
            slow.next = temp;
            temp = slow;
            slow = nextNode;
        }
        while(temp!=null){
            if(temp.val != head.val) return false;
            temp = temp.next;
            head = head.next;
        }
        return true;
    }
}

File: 234-palindrome-linked-list/palindrome-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
typedef struct ListNode node;
bool isPalindrome(struct ListNode* head){
    node *slow = head;
    node *fast = head;
    while (fast!=NULL && fast->next!=NULL){
        slow=slow->next;
        fast=fast->next->next;
    }
    node *temp=NULL;
    while(slow){
        node *p=slow->next;
        slow->next=temp;
        temp=slow;
        slow=p;
    }
    while(temp){
        if(temp->val!=head->val) return 0;
        temp=temp->next;
        head=head->next;
    }
    return 1;
}